Fix float exponents when writing a TDB file (TC ERROR DatabaseUtils)

rdamaral opened the pull request:
Hi, this is a small fix to '_print_Piecewise' function in the io.tdb module to improve compatibility with TC. I'm using this module to parse some of my tdbs and TC complains about float exponents (A constant integer must be used as the exponent in exponential TDB expressions). Thank you.

<!--

Thank you for pull request.

Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.

-->


Checklist
* [N/A] If any dependencies have changed, the changes are reflected in the
  * [N/A] `setup.py` (runtime requirements)
  * [N/A] `pyproject.toml` (build requirements)
  * [N/A] `requirements-dev.txt` (build and development requirements)

richardotis commented:
Only one test is failing, and I think it's safe for you to modify the pass condition for that test to match this new behavior, since it only differs in the amount of whitespace.

rdamaral commented:
Hi Richard, I’ve updated ‘test_database.py’, and the test should pass now. I’ve also generalized my previous rules to allow line breaks before any number addition/subtraction that isn’t part of an EXP or LN/LOG operation. The previous rules already accounted for that, but were overlooking some permissible break points.

richardotis commented on pycalphad/io/tdb.py:
```diff
@@ -541,7 +541,7 @@ def reflow_text(text, linewidth=80):
             while len(line) > linewidth:
                 linebreak_idx = linewidth - 1
-                while linebreak_idx > 0 and line[linebreak_idx] not in linebreak_chars:
+                while linebreak_idx > 0 and not re.match(r'[^E\(][+-][0-9]', line[linebreak_idx-1:linebreak_idx+2]) and line[linebreak_idx] not in linebreak_chars:
```
Can you clarify why this regex is needed?

richardotis replied:
See response in https://github.com/pycalphad/pycalphad/pull/516#issuecomment-2186652180

rdamaral commented:
Hi @richardotis. This regex basically matches any addition and subtraction of any number, ie. [+-][0-9], since they are not preceded by 'E' or '('. This should prevent breaking lines inside EXP or LN/LOG or power expression: '6.14599E-07', 'T**(-3)', and 'LOG(-3)', hence splitting these could lead to errors in TC. In practice, this makes the functions being breakable at the `linebreak_chars` plus at [+-][0-9], since they are not preceded by 'E' or '('.

I'll add a comment on the code about this implementation and merge it to https://github.com/pycalphad/pycalphad/pull/522

richardotis commented on pycalphad/io/tdb.py:
```diff
@@ -466,9 +466,19 @@ def _stringify_expr(self, expr):
+            #All the arguments in Mul should be tested and they're all combined to a single expression by ' * '
+            #    So we could stringify each argument as a list and join them together is ' * '
+            term_list = ['(' + self._stringify_expr(arg) + ')' if isinstance(arg,Add) else self._stringify_expr(arg) for arg in expr.args]
+            terms = ' * '.join(term_list)
```
Is it ok for the spaces in `' * '` to be there?

richardotis commented:
The test failures are related to https://github.com/pycalphad/pycalphad/pull/530 and can be ignored here.

Can you add a test for `reflow_text` to confirm the line-breaking behavior for floating-point exponents is correct now?

rdamaral commented:
Sure, I'll add it and let you know.

rdamaral commented:
Hi @richardotis, just added a new test ``test_reflow_text_for_line_breaks`` for testing the line breaks and updated ``test_tc_printer_exp`` for testing if floating-point exponents are being returned.
